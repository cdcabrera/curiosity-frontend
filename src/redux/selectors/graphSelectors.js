import { createSelector } from 'reselect';
import {
  rhsmConstants,
  RHSM_API_RESPONSE_TALLY_DATA_TYPES as TALLY_DATA_TYPES,
  RHSM_API_RESPONSE_TALLY_META_TYPES as TALLY_META_TYPES
} from '../../services/rhsm/rhsmConstants';

/**
 * Return a combined state, props object.
 *
 * @private
 * @param {object} state
 * @param {object} props
 * @param {object} data
 * @param {Array} data.metrics
 * @param {string} data.productId
 * @returns {object}
 */
const statePropsFilter = (state, props, { metrics = [], productId } = {}) => {
  const selectedMetrics = {};

  metrics.forEach(metricId => {
    selectedMetrics[metricId] = state.graph.tally?.[`${productId}_${metricId}`] || {};
  });

  return {
    ...selectedMetrics
  };
};

/**
 * Create selector, transform combined state, props into a consumable object.
 *
 * @type {{ metrics: object }}
 */
const selector = createSelector([statePropsFilter], response => {
  const metrics = response || {};
  const updatedResponseData = { pending: false, fulfilled: false, error: false, metrics: {} };
  const objEntries = Object.entries(metrics);
  let isPending = false;
  let isFulfilled = false;
  let errorCount = 0;

  objEntries.forEach(([metric, metricValue]) => {
    const { pending, cancelled, error, fulfilled, data: metricData, ...metricResponse } = metricValue;
    const { [rhsmConstants.RHSM_API_RESPONSE_DATA]: data = [], [rhsmConstants.RHSM_API_RESPONSE_META]: meta = {} } =
      metricData || {};
    const updatedPending = pending || cancelled || false;

    if (updatedPending) {
      isPending = true;
    }

    if (fulfilled) {
      isFulfilled = true;
    }

    if (error) {
      errorCount += 1;
    }

    updatedResponseData.metrics[metric] = {
      ...metricResponse,
      pending: updatedPending,
      error,
      fulfilled,
      data: data.map(
        (
          { [TALLY_DATA_TYPES.DATE]: date, [TALLY_DATA_TYPES.VALUE]: value, [TALLY_DATA_TYPES.HAS_DATA]: hasData },
          index
        ) => ({
          x: index,
          y: value,
          date,
          hasData
        })
      ),
      meta: {
        count: meta[TALLY_META_TYPES.COUNT] || 0,
        metricId: meta[TALLY_META_TYPES.METRIC_ID],
        productId: meta[TALLY_META_TYPES.PRODUCT],
        totalMonthly: {
          date: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.DATE],
          hasData: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.HAS_DATA],
          value: meta[TALLY_META_TYPES.TOTAL_MONTHLY]?.[TALLY_META_TYPES.VALUE]
        }
      }
    };
  });

  if (errorCount === objEntries.length) {
    updatedResponseData.error = true;
  } else if (isPending) {
    updatedResponseData.pending = true;
  } else if (isFulfilled) {
    updatedResponseData.fulfilled = true;
  }

  return updatedResponseData;
});

/**
 * Expose selector instance. For scenarios where a selector is reused across component instances.
 *
 * @param {object} data Pass additional data.
 * @returns {{metrics: object}}
 */
const makeSelector = data => (state, props) => ({
  ...selector(state, props, data)
});

const graphSelectors = {
  graph: selector,
  makeGraph: makeSelector
};

export { graphSelectors as default, graphSelectors, selector, makeSelector };
