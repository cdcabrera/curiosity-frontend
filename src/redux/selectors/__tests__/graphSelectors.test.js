import graphSelectors from '../graphSelectors';
import {
  RHSM_API_RESPONSE_TALLY_META_TYPES as TALLY_META_TYPES,
  rhsmConstants
} from '../../../services/rhsm/rhsmConstants';

describe('GraphSelectors', () => {
  it('should return specific selectors', () => {
    expect(graphSelectors).toMatchSnapshot('selectors');
  });

  it('should pass minimal data on missing a reducer response', () => {
    const state = {};
    expect(graphSelectors.graph(state)).toMatchSnapshot('missing reducer error');
  });

  it('should handle an error state', () => {
    const data = {
      productId: 'Lorem Ipsum',
      metrics: ['Dolor Sit', 'Et all']
    };
    const props = {};
    const state = {
      graph: {
        tally: {
          'Lorem Ipsum_Dolor Sit': {
            error: true,
            errorMessage: `I'm a teapot`,
            fulfilled: false,
            pending: false,
            status: 418,
            meta: {},
            metaId: 'Lorem Ipsum_Dolor Sit',
            metaIdMetric: { id: 'Lorem Ipsum', metric: 'Dolor Sit' },
            metaQuery: {}
          },
          'Lorem Ipsum_Et all': {
            error: true,
            errorMessage: `I'm a teapot`,
            fulfilled: false,
            pending: false,
            status: 418,
            meta: {},
            metaId: 'Lorem Ipsum_Et all',
            metaIdMetric: { id: 'Lorem Ipsum', metric: 'Et all' },
            metaQuery: {}
          }
        }
      }
    };

    expect(graphSelectors.graph(state, props, data)).toMatchSnapshot('error state');
  });

  it('should handle a cancelled state', () => {
    const data = {
      productId: 'Lorem Ipsum',
      metrics: ['Dolor Sit']
    };
    const props = {};
    const state = {
      graph: {
        tally: {
          'Lorem Ipsum_Dolor Sit': {
            date: 'mock date',
            cancelled: true,
            error: false,
            errorMessage: '',
            fulfilled: false,
            pending: false,
            meta: {},
            metaId: 'Lorem Ipsum_Dolor Sit',
            metaIdMetric: { id: 'Lorem Ipsum', metric: 'Dolor Sit' },
            metaQuery: {}
          }
        }
      }
    };

    expect(graphSelectors.graph(state, props, data)).toMatchSnapshot('cancelled state');
  });

  it('should handle a pending state', () => {
    const data = {
      productId: 'Lorem Ipsum',
      metrics: ['Dolor Sit']
    };
    const props = {};
    const state = {
      graph: {
        tally: {
          'Lorem Ipsum_Dolor Sit': {
            error: false,
            errorMessage: '',
            fulfilled: false,
            pending: true,
            meta: {},
            metaId: 'Lorem Ipsum_Dolor Sit',
            metaIdMetric: { id: 'Lorem Ipsum', metric: 'Dolor Sit' },
            metaQuery: {}
          }
        }
      }
    };

    expect(graphSelectors.graph(state, props, data)).toMatchSnapshot('pending state');
  });

  it('should handle a fulfilled state', () => {
    const data = {
      productId: 'Lorem Ipsum',
      metrics: ['Dolor Sit']
    };
    const props = {};
    const state = {
      graph: {
        tally: {
          'Lorem Ipsum_Dolor Sit': {
            error: false,
            errorMessage: '',
            fulfilled: true,
            pending: false,
            meta: {},
            metaId: 'Lorem Ipsum_Dolor Sit',
            metaIdMetric: { id: 'Lorem Ipsum', metric: 'Dolor Sit' },
            metaQuery: {},
            date: '2019-09-05T00:00:00.000Z',
            data: {
              [rhsmConstants.RHSM_API_RESPONSE_DATA]: [
                {
                  [rhsmConstants.RHSM_API_RESPONSE_TALLY_DATA_TYPES.DATE]: '2019-09-04T00:00:00.000Z',
                  [rhsmConstants.RHSM_API_RESPONSE_TALLY_DATA_TYPES.VALUE]: 10,
                  [rhsmConstants.RHSM_API_RESPONSE_TALLY_DATA_TYPES.HAS_DATA]: true
                },
                {
                  [rhsmConstants.RHSM_API_RESPONSE_TALLY_DATA_TYPES.DATE]: '2019-09-05T00:00:00.000Z',
                  [rhsmConstants.RHSM_API_RESPONSE_TALLY_DATA_TYPES.VALUE]: 15,
                  [rhsmConstants.RHSM_API_RESPONSE_TALLY_DATA_TYPES.HAS_DATA]: false
                }
              ],
              [rhsmConstants.RHSM_API_RESPONSE_META]: {
                [TALLY_META_TYPES.COUNT]: 2,
                [TALLY_META_TYPES.METRIC_ID]: 'Dolor Sit',
                [TALLY_META_TYPES.PRODUCT]: 'Lorem Ipsum',
                [TALLY_META_TYPES.TOTAL_MONTHLY]: {
                  [TALLY_META_TYPES.DATE]: '2019-09-05T00:00:00.000Z',
                  [TALLY_META_TYPES.HAS_DATA]: true,
                  [TALLY_META_TYPES.VALUE]: 25
                }
              }
            },
            status: 304
          }
        }
      }
    };

    expect(graphSelectors.graph(state, props, data)).toMatchSnapshot('fulfilled state');
  });
});
