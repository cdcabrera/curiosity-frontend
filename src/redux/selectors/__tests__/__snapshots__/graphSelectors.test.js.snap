// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphSelectors should handle a cancelled state: cancelled state 1`] = `
Object {
  "error": false,
  "fulfilled": false,
  "metrics": Object {
    "Dolor Sit": Object {
      "data": Array [],
      "date": "mock date",
      "error": false,
      "errorMessage": "",
      "fulfilled": false,
      "meta": Object {
        "count": 0,
        "metricId": undefined,
        "productId": undefined,
        "totalMonthly": Object {
          "date": undefined,
          "hasData": undefined,
          "value": undefined,
        },
      },
      "metaId": "Lorem Ipsum_Dolor Sit",
      "metaIdMetric": Object {
        "id": "Lorem Ipsum",
        "metric": "Dolor Sit",
      },
      "metaQuery": Object {},
      "pending": true,
    },
  },
  "pending": true,
}
`;

exports[`GraphSelectors should handle a fulfilled state: fulfilled state 1`] = `
Object {
  "error": false,
  "fulfilled": true,
  "metrics": Object {
    "Dolor Sit": Object {
      "data": Array [
        Object {
          "date": "2019-09-04T00:00:00.000Z",
          "hasData": true,
          "x": 0,
          "y": 10,
        },
        Object {
          "date": "2019-09-05T00:00:00.000Z",
          "hasData": false,
          "x": 1,
          "y": 15,
        },
      ],
      "date": "2019-09-05T00:00:00.000Z",
      "error": false,
      "errorMessage": "",
      "fulfilled": true,
      "meta": Object {
        "count": 2,
        "metricId": "Dolor Sit",
        "productId": "Lorem Ipsum",
        "totalMonthly": Object {
          "date": "2019-09-05T00:00:00.000Z",
          "hasData": true,
          "value": 25,
        },
      },
      "metaId": "Lorem Ipsum_Dolor Sit",
      "metaIdMetric": Object {
        "id": "Lorem Ipsum",
        "metric": "Dolor Sit",
      },
      "metaQuery": Object {},
      "pending": false,
      "status": 304,
    },
  },
  "pending": false,
}
`;

exports[`GraphSelectors should handle a pending state: pending state 1`] = `
Object {
  "error": false,
  "fulfilled": false,
  "metrics": Object {
    "Dolor Sit": Object {
      "data": Array [],
      "error": false,
      "errorMessage": "",
      "fulfilled": false,
      "meta": Object {
        "count": 0,
        "metricId": undefined,
        "productId": undefined,
        "totalMonthly": Object {
          "date": undefined,
          "hasData": undefined,
          "value": undefined,
        },
      },
      "metaId": "Lorem Ipsum_Dolor Sit",
      "metaIdMetric": Object {
        "id": "Lorem Ipsum",
        "metric": "Dolor Sit",
      },
      "metaQuery": Object {},
      "pending": true,
    },
  },
  "pending": true,
}
`;

exports[`GraphSelectors should handle an error state: error state 1`] = `
Object {
  "error": true,
  "fulfilled": false,
  "metrics": Object {
    "Dolor Sit": Object {
      "data": Array [],
      "error": true,
      "errorMessage": "I'm a teapot",
      "fulfilled": false,
      "meta": Object {
        "count": 0,
        "metricId": undefined,
        "productId": undefined,
        "totalMonthly": Object {
          "date": undefined,
          "hasData": undefined,
          "value": undefined,
        },
      },
      "metaId": "Lorem Ipsum_Dolor Sit",
      "metaIdMetric": Object {
        "id": "Lorem Ipsum",
        "metric": "Dolor Sit",
      },
      "metaQuery": Object {},
      "pending": false,
      "status": 418,
    },
    "Et all": Object {
      "data": Array [],
      "error": true,
      "errorMessage": "I'm a teapot",
      "fulfilled": false,
      "meta": Object {
        "count": 0,
        "metricId": undefined,
        "productId": undefined,
        "totalMonthly": Object {
          "date": undefined,
          "hasData": undefined,
          "value": undefined,
        },
      },
      "metaId": "Lorem Ipsum_Et all",
      "metaIdMetric": Object {
        "id": "Lorem Ipsum",
        "metric": "Et all",
      },
      "metaQuery": Object {},
      "pending": false,
      "status": 418,
    },
  },
  "pending": false,
}
`;

exports[`GraphSelectors should pass minimal data on missing a reducer response: missing reducer error 1`] = `
Object {
  "error": true,
  "fulfilled": false,
  "metrics": Object {},
  "pending": false,
}
`;

exports[`GraphSelectors should return specific selectors: selectors 1`] = `
Object {
  "graph": [Function],
  "makeGraph": [Function],
}
`;
